documentation/general/dotnet-run-file-multi-file.md

```diff
@@ -0,0 +1,136 @@
+# Multi-file support in `dotnet run app.cs`
+
+## Scenarios
+
+Some multi-file scenarios to consider:
+
+### Direcotry-based apps
+
+- Directory of files that are all part of the same app
+- Only a single entry-point file
+- Might be using differnt SDKs, e.g. `Microsoft.NET.Sdk` vs. `Microsoft.NET.Sdk.Web`, that have different default items
+
+Example console app:
+
+```text
+- myapp/
+  - myapp.cs
+  - helpers.cs
+```
+
+Example Web API app:
+
+```text
+- webapi/
+  - webapi.cs
+  - helpers.cs
+  - customerapi.cs
+  - productapi.cs
+  - appsettings.json
+```
+
+Example Razor app:
+
+```text
+- razorapp/
+  - razorapp.cs
+  - helpers.cs
+  - Pages/
+    - _ViewImports.razor
+    - Index.razor
+    - App.razor
+    - Layout.razor
+  - appsettings.json
+```
+
+Example Aspire app host app:
+
+```text
+- apphost/
+  - apphost.cs
+  - helpers.cs
+  - appsettings.json
+```
+
+### Directories with multiple separate entry-point files
+
+- Each entry-point file is a separate app
+- Some entry-point files require types from some non-entry-point files
+- Entry-point files don't require types from other entry-point files
+
+Example:
+
+```text
+- apps/
+  - app1.cs
+  - app2.cs
+  - webapi1.cs
+  - helpers.cs
+```
```

Comment:
Curious: do you have any concrete examples of an app that you see built this way?  The one case I can definitely see is infra. Basically I see dirs today with a bunch of `test-blah.ps1`, `build-tests.ps1`, etc ... that have a simple lift and shift to `.cs` entry point files. It's harder for me to see the case in a web app.

Reply:
Btw, to bring back support for this, I think we need a change in Roslyn so it's possible to pass (via a feature flag for example) the entry-point file path. Then Roslyn itself excludes other entry-point files from the compilation. The alternative is detecting entry points in SDK itself but that's problematic because we cannot 100% correctly parse files when we don't have access to DefineConstants. Imagine a file that has top-level statements under `#if SOMETHING` and there is `<DefineConstants>SOMETHING</>` in a Directory.Build.props. The SDK cannot know that when deciding which files to pass down to the compilation, only csc.exe knows the set of effective DefineConstants. Perhaps this inconsistency would be fine for starters if we don't want to change Roslyn like that though.

Reply:
It's a bit of a catch-22 though -- csc.exe doesn't _really_ know the DefineConstants either -- those just come from MSBuild. Which is only known once we've already created the project. Imagine the truly evil case of your Directory.Build.props making the DefineProps conditional on the existence of a Compile item! We obviously don't care about such a scenario, but fundamentally we'll have to decide how we break the dependency loop of:

1. To build the project file, we have to know entry point files.
2. To know entry point files, we have to parse files.
3. To parse files, we have to know the compiler settings.
4. To know the compiler settings, we must build the project file.

Reply:
Ah, so csc might get file paths that ultimately get ignored later?

Reply:
Yes, because the logic of which file has an entry point is complicated (especially with different variants of Main) and should be left to the compiler to decide. So it needs to know all paths and then decide which to ignore. It might not even ignore the other entry point *files* completely, just decide which entry point to set in the dll as Main (so classes from the other entry point files are still usable for example).

Reply:
> It might not even ignore the other entry point files completely, just decide which entry point to set in the dll as Main (so classes from the other entry point files are still usable for example).

This is where I'd love to land.